### ISSUE TITLE
[bug] Memory leaks in default.qubit.jax, pulse module or pulse gradient transform

### ISSUE THREAD

**Issue by Qottmann** [PennyLaneAI/pennylane#4264]
Repeated execution of pulse programing circuits anomalously builds up memory.
It is unclear if that is happening in the pulse gradient transform, somewhere in the pulse module or in the `default.qubit.jax` device.

You can reproduce the behavior by letting this script run for a while and note how your computer's memory increases with every iteration (so some unreferenced values are not garbage collected) by looking at the table of processes (`top` on linux)

```python
n_wires = 2
atol = 1e-8
timespan = 0.1
H_obj = qml.Hamiltonian(
    [1. for i in range(n_wires)],
    [qml.PauliZ(i) @ qml.PauliZ((i+1)%n_wires) for i in range(n_wires)]
)
H = qml.dot(
    [qml.pulse.constant for i in range(n_wires)],
    [qml.PauliY(i) @ qml.PauliY((i+1)%n_wires) for i in range(n_wires)]
)

dev = qml.device("default.qubit.jax", wires=n_wires)

def circuit(params):
    qml.evolve(H, atol=atol)(params, t=timespan)
    return qml.expval(H_obj)


num_split_times = 6

cost_ps  = qml.QNode(
    circuit,
    dev, 
    interface="jax", 
    diff_method=qml.gradients.stoch_pulse_grad, 
    num_split_times=num_split_times, 
    use_broadcasting=False
)

params = jnp.ones(n_wires)

for i in range(200):
    print(f"{i} / {200}")
    grad = jax.grad(cost_ps)(params)
```

**Comment by Qottmann** [comment]
Note that this most likely lead to CI runners not finishing and we need to hot-fix it by explicitly emptying jax caches [here](https://github.com/PennyLaneAI/pennylane/pull/4215/commits/56ba598c069bbb0e4254ceb68ee5e6d8387679f5).

Once the system reaches `100%MEM` it does not yield an OOM error but just crashed the python kernel, hence the ambiguity of the failed CI runners above.

**Comment by albi3ro** [comment]
The problem also exists with `DefaultQubit2`, so its not purely coming from the device.

**Comment by Qottmann** [comment]
Did you run the same circuit, i.e. with the `ParametrizedEvolution` and `qml.gradients.stoch_pulse_grad` transform - or for a "regular" circuit?

**Comment by albi3ro** [comment]
Still crashed my kernel after about 8 iterations.

**Comment by mlxd** [comment]
Had a quick look into this --- I don't think this is a memory leak, and I think it is the internal JAX cache holding onto all compiled methods it encounters. The issue is that PL doesn't easily allow JAX functions to be compiled once and resued --- due to some representation issues with how we define the interfaces (others may know better than I), it freshly compiles everything it sees on every pass.

For the latest JAX, if you want to clear the compilation cache between runs, and cap the memory wall (overall around 500 MB on my machine for the first example provided), you can add `jax.clear_caches()` ([docs](https://jax.readthedocs.io/en/latest/_autosummary/jax.clear_caches.html) and [discussion](https://github.com/google/jax/discussions/10826) )after the gradient in the loop. Not sure if it is worth using JAX if we do this, but it at least keeps the memory boundary capped on my machine (AMD64, Linux).

**Comment by Qottmann** [comment]
Thanks @mlxd !

I think I figured out the issue. It is that we define a new `fun` in https://github.com/PennyLaneAI/pennylane/blob/f6de3e61831782334d807148218d2d426726cc9e/pennylane/pulse/parametrized_evolution.py#L493 at every call and jax does not realize it is the same function, and recompiles because it is a static argument of `odeint`.

This snippet should exemplify the problem. In the first case I dont re-define fun and the caches is unaltered
```python
def fun(y, t):
    return y * t

def bigfun(fun, y, t):
    return fun(t, t)

y, t = jnp.array(0.5), jnp.array(0.4)

bigfun = jax.jit(bigfun, static_argnums=0)

bigfun(fun, y, t)

print(f"cache: ", bigfun._cache_size())

bigfun(fun, y, t)

print(f"cache: ", bigfun._cache_size())
```
```pycon
cache:  1
cache:  1
```

But if I re-define the exact same fun before, it gets compiled again:
```python
def fun(y, t):
    return y * t

def bigfun(fun, y, t):
    return fun(t, t)

y, t = jnp.array(0.5), jnp.array(0.4)

bigfun = jax.jit(bigfun, static_argnums=0)

bigfun(fun, y, t)

print(f"cache: ", bigfun._cache_size())

def fun(y, t):
    return y * t

bigfun(fun, y, t)

print(f"cache: ", bigfun._cache_size())
```
```pycon
cache:  1
cache:  2
```

So we are internally caching a new function for every call, even though we are using the same functions.

**Comment by Qottmann** [comment]
Vendoring `jax.experimental.odeint` into pennylane is something we cant commit to, hence we leave this issue as "wont fix" for now.

Note that this problem currently only occurs when the qnode / overall execution is not jitted. This is the case in particular for classically simulating the hardware gradient methods `stoch_pulse_grad` and `pulse_generator`. Unless you investigate these particular gradient estimation methods, we recommend using backpropagation (default for `jax.grad(qnode)` if not otherwise set in the qnode definition).
